lib/apollo-client.ts

```diff
@@ -255,5 +258,53 @@ export function initClient({ initialState, twoFactorAuthContext }: any = {}): Re
     apolloClient = createClient({ initialState, twoFactorAuthContext });
   }
 
+  // If the page has Next.js data fetching methods that use Apollo Client, the initial state
+  // get hydrated here
+  if (initialState) {
+    // Get existing cache, loaded during client side data fetching
+    const existingCache = apolloClient.extract();
+
+    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
+    const data = mergeDeep(initialState, existingCache);
+
+    // Restore the cache with the merged data
+    apolloClient.cache.restore(data);
+  }
+
   return apolloClient;
 }
+
+/**
+ * A helper to easily plug Apollo on functional components that use `getServerSideProps` thats make sure that
+ * the server-side query and the client-side query/variables are the same; to properly rehydrate the cache.
+ */
+export function getSSRQueryHelpers({
```

Comment:
I really like this abstraction, I thought about doing something like this in the first implementation but couldn't figure out exactly what I wanted to do.

We should implement type support to tie `context.query` to `props`, `variables`, and the functions that we need to pass.